xapian-core/cluster/cluster.cc

```diff
@@ -0,0 +1,586 @@
+	values[term] = wt;
+	magnitude += wt * wt;
+    }
+}
+
+Centroid::Centroid(Point &p) {
+    LOGCALL_CTOR(API, "Centroid", p);
+    for (TermIterator it = p.termlist_begin(); it != p.termlist_end(); ++it)
+	values[*it] = p.get_value(*it);
+    magnitude = p.get_magnitude();
+}
+
+void
+Centroid::divide(double num)
+{
+    LOGCALL_VOID(API, "Centroid::divide", num);
+    unordered_map<string, double>::iterator it;
+    for (it = values.begin(); it != values.end(); ++it)
+	it->second /= num;
+}
+
+void
+Centroid::clear()
+{
+    LOGCALL_VOID(API, "Centroid::clear", NO_ARGS);
+    values.clear();
+}
+
+void
+Centroid::recalc_magnitude()
+{
+    LOGCALL_VOID(API, "Centroid::recalc_magnitude", NO_ARGS);
+    magnitude = 0;
+    unordered_map<string, double>::iterator it;
+    for (it = values.begin(); it != values.end(); ++it)
+	magnitude += it->second * it->second;
+}
+
+Cluster&
+Cluster::operator=(const Cluster &other)
+{
+    // pointers are reference counted.
+    internal = other.internal;
+    return *this;
+}
+
+Cluster::Cluster(const Cluster &other)
+    : internal(other.internal)
+{
+}
+
+Cluster::Cluster() : internal(new Xapian::Cluster::Internal)
+{
+    LOGCALL_CTOR(API, "Cluster", NO_ARGS);
+}
+
+Cluster::Cluster(const Centroid &centroid)
+    : internal(new Xapian::Cluster::Internal(centroid))
+{
+    LOGCALL_CTOR(API, "Cluster", centroid);
+}
+
+Cluster::~Cluster()
+{
+    LOGCALL_DTOR(API, "Cluster");
+}
+
+Centroid::Centroid()
+{
+    LOGCALL_CTOR(API, "Centroid", NO_ARGS);
+}
+
+DocumentSet
+Cluster::get_documents()
+{
+    LOGCALL(API, DocumentSet, "Cluster::get_documents", NO_ARGS);
+    return internal->get_documents();
+}
+
+DocumentSet
+Cluster::Internal::get_documents()
+{
+    DocumentSet docs;
+    for (auto&& point : cluster_docs)
+	docs.add_document(point.get_document());
+    return docs;
+}
+
+Point
+Cluster::get_index(unsigned int index) const
+{
+    LOGCALL(API, Point, "Cluster::get_index", index);
+    return internal->get_index(index);
+}
+
+Point
+Cluster::Internal::get_index(unsigned int index) const
+{
+    return cluster_docs[index];
+}
+
+ClusterSet&
+ClusterSet::operator=(const ClusterSet &other)
+{
+    // pointers are reference counted.
+    internal = other.internal;
+    return *this;
+}
+
+ClusterSet::ClusterSet(const ClusterSet &other)
+    : internal(other.internal)
+{
+}
+
+ClusterSet::ClusterSet() : internal(new Xapian::ClusterSet::Internal)
+{
+}
+
+ClusterSet::~ClusterSet()
+{
+}
+
+doccount
+ClusterSet::Internal::size() const
+{
+    return clusters.size();
+}
+
+doccount
+ClusterSet::size() const
+{
+    LOGCALL(API, doccount, "ClusterSet::size", NO_ARGS);
+    return internal->size();
+}
+
+void
+ClusterSet::Internal::add_cluster(const Cluster &c)
+{
+    clusters.push_back(c);
+}
+
+void
+ClusterSet::add_cluster(const Cluster &c)
+{
+    LOGCALL_VOID(API, "ClusterSet::add_cluster", c);
+    internal->add_cluster(c);
+}
+
+Cluster&
+ClusterSet::Internal::get_cluster(doccount i)
```

Comment:
If that doesn't break how this is use, definitely.

If we need to provide access to a non-const reference, then it's usual to provide both a const and non-const versions of `operator[]` (which return correspondingly const and non-const references), and that can also happen with the `get_cluster()` internal methods which underlie them.  The compiler will pick the appropriate one in each case (preferring the const version where possible).